### DISCUSSION TITLE
Dependency Injection with new Hosting Model

### DISCUSSION THREAD

**Discussion by shawndewet** [akkadotnet/akka.net#6066] [Q&A] [answered]
I have a Blazor server app into which I am registering Akka.Net as follows:
```

builder.Services.AddSingleton<SignalRConnectionList>();
builder.Services.AddRazorPages();
builder.Services.AddServerSideBlazor();
builder.Services.AddAkka("AkkaESPoC", (configurationBuilder, provider) =>
{
    configurationBuilder
        .WithRemoting(hostName, port)
        .WithClustering(new ClusterOptions()
        { Roles = new[] { "Web" }, SeedNodes = seeds })
        .WithActors((system, registry) =>
        {
            var viewActorProps = system.ActorOf(Props.Create<QuestViewActor>(provider), "questview"); 
        });
});
```

Note how I am injecting the IServiceProvider into Actor props, and the actor's constructor creates a scope as follows:

```
public QuestViewActor(IServiceProvider serviceProvider)
        {
            _scope = serviceProvider.CreateScope();
... other code
        }
```

However, in the Receive method of the Actor, the services that are resolved are NOT the same services that are resolved elsewhere in the application:
```
Receive<QuestState>(async message =>
            {
                var _questHub = _scope.ServiceProvider.GetRequiredService<IHubContext<QuestHub>>();
                await _questHub.Clients.All.SendAsync("AddQuest", message.Data);

                var counter = _scope.ServiceProvider.GetRequiredService<SignalRConnectionList>();
                _log.Info($"Have {counter.Count} connections");
         });
```

I'm trying to resolve a SignalR `IHubContext` above, but the hub there has zero connections. So no views are receiving the messages.
I have confirmed that the scope inside the Actor Receive method is resolving different instances of services, in that the `SignalRConnectionList` is registered as a *singleton*, and elsewhere in my app, that .Count returns a value equal to the connections,  BUT, in the above Actor Receive method, that count is zero. 

What am I missing?  How do get my actor to use the same DI Service Provider as the rest of my Blazor Application?

**Comment by bauersystems** [comment]
I may be wrong, as I am only new to Akka.net myself; however, I believe the issue may be to do with you attempting to pull a singleton via a scope. Instead of using the scope, try resolving the SignalRConnectionList directly from the Service Provider.

The other thing I did for mine was to not inject the IServiceProvider and use IDependencyResolver (from Akka.DependencyInjection) instead. Here is an example using IHttpClientFactory:
`IDependencyResolver resolver = DependencyResolver.For(Context.System).Resolver;
_httpClientFactory = resolver.GetService<IHttpClientFactory>();`

If you need to pull in a scoped dependency, you can create a scope from the Dependency Resolver like the following:
`IDependencyResolver resolver = DependencyResolver.For(Context.System).Resolver;
resolver.CreateScope();`
It seems to act the same as creating one through IServiceProvider, so I assume it is a wrapper under the surface (again, someone more experienced than me can confirm).
One thing to keep in mind when using a scope is to manage it in the PreStart & PostStart (as per the documentation https://getakka.net/articles/actors/dependency-injection.html#managing-lifecycle-dependencies-with-akkadependencyinjection)

**Reply by shawndewet** [reply]
Sadly I have tried both of the other approaches as well, and they both yielded the same results.  But thanks very much for your input.

Here is the full Actor Receive method, indicating the commented-out approaches:

```
Receive<QuestState>(async message =>
            {
                 //using DependencyResolver
                //var _questHub = Akka.DependencyInjection.DependencyResolver.For(Context.System).Resolver.GetService<IHubContext<QuestHub>>();
                //var counter = Akka.DependencyInjection.DependencyResolver.For(Context.System).Resolver.GetService<SignalRConnectionList>();

                //using injected IServiceProvider
                //var _questHub = _serviceProvider.GetRequiredService<IHubContext<QuestHub>>();
                //var counter = _serviceProvider.GetRequiredService<SignalRConnectionList>();

                //using _scope
                var _questHub = _scope.ServiceProvider.GetRequiredService<IHubContext<QuestHub>>();
                var counter = _scope.ServiceProvider.GetRequiredService<SignalRConnectionList>();

                _log.Info($"Have {counter.Count} connections");
                await _questHub.Clients.All.SendAsync("AddQuest", message.Data);
            });
```

In all three cases, the log line shows a counter with zero connections.  While the same Counter injected elsewhere in the Blazor app shows the expected count of SignalR connections.

Tell me though, @bauersystems, are you bootstrapping your system with the `IHostedService` approach, or with the `builder.Services.AddAkka()` approach?  If it's the former, I suspect there may be a bug in the `.AddAkka()` approach that I am using.

**Reply by shawndewet** [reply]
I've even tried this approach of using DI in to boostrapping of the Akka system:
```
.WithActors((system, registry) =>
        {
            var resolvedProps = Akka.DependencyInjection.DependencyResolver.For(system).Props<QuestViewActor>();
            system.ActorOf(resolverProps, "questview");
        });
```

It correctly instantiates the actor, and injects the IServiceProvider dependency.  But still the same result.  I'm feeling quite disheartened at this point.

**Reply by shawndewet** [reply]
I just can't find a working example of Akka.net with SignalR, using the new `services.AddAkka()` bootstrapping approach.

**Reply by bauersystems** [reply]
I am using the `services. AddAkka()` approach. I'm not a fan of HOCON, I prefer to configure things programmatically.

I'm curious why you are using clustering? Why not take that out and see if you can get it to work on a standalone ActorSystem first?

Also, is this a work sensitive project or a project just to learn on? (As in, are you able to make your repository public and share the code?).

**Reply by shawndewet** [reply]
@bauersystems Uhg I'm with you regarding your dislike of Hocon.  The .AddAkka() tooling is a much better approach.

Clustering is in there because it's part of a larger PoC that I am building, based on this example: https://github.com/petabridge/akkadotnet-code-samples/tree/master/src/clustering/sharding-sqlserver

You can see my repo here: https://github.com/shawndewet/AkkaESPoC  The ReadMe is updated with the current state of the project.

As indicated in the ReadMe, Distributed PubSub is working, but the SignalR message is where I am struggling.

**Reply by shawndewet** [reply]
As per your suggestion I've also configured a simple single-project Blazor app with Akka and SignalR, and am frustrated to say that it works!  See here: https://github.com/shawndewet/BlazorWithAkkaSignalRSln.

**Reply by bauersystems** [reply]
I am not sure what is happening with your AkkaESPoC project, it is getting beyond my capabilities (as I mentioned previously, I am still learning too).

I did notice that you are attempting a lot of concepts at the same time (Clustering, Persistence, Pub/Sub, Protobuf). It might be worth using your BlazorWithAkkaSignalRSln project as a testing ground, adding new concepts to it one at a time to see how each concept works individually; then once you are finished you can have another look at AkkaESPoC and apply what you have learnt.

On a side note, for serialization I was running into trouble with the standard JSON serializer not being able to deal with my messages correctly, and Protobuf takes quite a bit to setup correctly (as you have seen), so I tried the new Hyperion one and it worked very well. To set it up with Akka.Hosting, it is as simple as `.WithCustomSerializer("hyperion", new[] { typeof(object) }, system => new HyperionSerializer(system))`.

I know that Aaron talks about ideally using Protobuf, however in the beginning it might be worth making the learning curve easier for yourself and just using a simpler serializer.

**Reply by shawndewet** [reply]
Thanks for your input @bauersystems.
I took a couple of hours yesterday and did exactly that...built a new PoC adding in concepts one at a time.
The core issue turned out to be frickin' nuget references!!!!

Because I was building the PoC based on the sharding-sqlserver sample I referred to earlier, I had the same `Akka.Persistence.SqlServer.Hosting` nuget reference in my Blazor project as that sample had in its web project.

So I replaced that with the references that were being used in my from-scratch BlazorWithAkkaSignalRSln project, and suddenly the issue was resolved!  Yay.
So it must be that `Akka.Persistence.SqlServer.Hosting` is referencing earlier versions of the core Akka.Hosting dll's that do not yet do the DI stuff correctly.

But yay, the problem is solved!  The actors in the Blazor project are now correctly doing SignalR!

**Reply by shawndewet** [reply]
Oh, and thanks for the tip about Hyperion @bauersystems !  Much much simpler!

**Reply by bauersystems** [reply]
I am glad to hear that you got it sorted out.

You might want to mark the discussion as answered ;-)

**Comment by shawndewet** [comment, marked as answer]
Thanks for your input @bauersystems.
I took a couple of hours yesterday and did exactly that...built a new PoC adding in concepts one at a time.
The core issue turned out to be frickin' nuget references!!!!

Because I was building the PoC based on the sharding-sqlserver sample I referred to earlier, I had the same Akka.Persistence.SqlServer.Hosting nuget reference in my Blazor project as that sample had in its web project.

So I replaced that with the references that were being used in my from-scratch BlazorWithAkkaSignalRSln project, and suddenly the issue was resolved! Yay.
So it must be that Akka.Persistence.SqlServer.Hosting is referencing earlier versions of the core Akka.Hosting dll's that do not yet do the DI stuff correctly.

But yay, the problem is solved! The actors in the Blazor project are now correctly doing SignalR!
